pkg/mount: optimizations

kolyshkin opened the pull request:
Various minor things here and there to streamline pkg/mount operations.

kolyshkin committed:
pkg/mount: Mount: minor optimization

Eliminate double call to parseOptions() from Mount()

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
daemon/mountVolumes(): eliminate MakeRPrivate call

It is sufficient to add "rprivate" to mount flags.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented on pkg/mount/mount.go:
```diff
@@ -102,21 +102,12 @@ func Mounted(mountpoint string) (bool, error) {
-// *regardless* if the target path is not already mounted. Options must be
-// specified like the mount or fstab unix commands: "opt1=val1,opt2=val2". See
-// flags.go for supported option flags.
-func ForceMount(device, target, mType, options string) error {
```
I see there's a couple of uses of this function in the wild; https://github.com/search?l=Go&q=%22mount.ForceMount%28%22&type=Code

Should we keep the wrapper and add a `deprecated` to the doc?

```
// ForceMount will mount a filesystem according to the specified configuration,
// *regardless* if the target path is not already mounted. Options must be
// specified like the mount or fstab unix commands: "opt1=val1,opt2=val2". See
// flags.go for supported option flags.
// Deprecated: use Mount() instead
```

kolyshkin replied:
I should have checked if there are any external users... My new commits leave this function as it was.

thaJeztah commented on pkg/mount/sharedsubtree_linux.go:
```diff
@@ -67,5 +67,5 @@ func ensureMountedAs(mountPoint, options string) error {
 	}
 
-	return ForceMount("", mountPoint, "none", options)
+	return mount("", mountPoint, uintptr(0), options)
```
This is failing;

```
00:41:38 pkg/mount/sharedsubtree_linux.go:70:31:warning: cannot use uintptr(0) (constant 0 of type uintptr) as string value in argument to mount (gosimple)
00:41:38 pkg/mount/sharedsubtree_linux.go:70:50:warning: too few arguments in call to mount (unconvert)
00:41:38 pkg/mount/sharedsubtree_linux.go:70:43:warning: cannot use options (variable of type string) as uintptr value in argument to mount (unconvert)
00:41:38 pkg/mount/sharedsubtree_linux.go:70:31:warning: cannot use uintptr(0) (constant 0 of type uintptr) as string value in argument to mount (unconvert)
```

kolyshkin replied:
Thanks, I was too soon to push this. Please see the new commits.

kolyshkin committed:
pkg/mount: MakeMount: minor optimization

Current code in MakeMount parses /proc/self/mountinfo twice:
first in call to Mounted(), then in call to Mount(). Use
ForceMount() to eliminate such double parsing.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
OK, let the `ForceMount` stay as it is... A new version pushed doing same optimizations without removing `ForceMount`.

kolyshkin commented:
OK it's too late now, need to take a fresh look tomorrow; setting WIP for now

kolyshkin commented:
Some night thoughs:

1. `MakeMount` (I blame myself for choosing such a bad name! should have been `MakeBindMount`) does essentially the same thing as Mount()
2. ensureMountedAs does the same thing as Mount(), too

Need to carefully look at it and minimize the code flow.

kolyshkin committed:
TestContainersAPICreateMountsCreate: minor optimization

Don't use two-stage mount in TestContainersAPICreateMountsCreate();
apparently it was written before mount.Mount() could accept propagation
flags.

While at it, remove rw as this is the default.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
pkg/mount/Make*: optimize

The only option we supply is either BIND or a mount propagation flag,
so it makes sense to specify the flag value directly, rather than using
parseOptions() every time.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
daemon/mountVolumes: no need to specify fstype

For bind mounts, fstype argument to mount(2) is ignored.
Usual convention is either empty string or "none".

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
OK I think this one is good to go. @cpuguy83 PTAL

thaJeztah reviewed (approved):
SGTM

ping @cpuguy83

thaJeztah commented:
ping @cpuguy83 PTAL

cpuguy83 reviewed (approved):
LGTM

olljanat commented:
Derek add label: rebuild/*
